File: packages/sim/src/movementKernel.ts
import { idiv, vecLen } from './fixed.js';
import { isWalkableCell } from './grid.js';
import { GRID_H, GRID_W } from './constants.js';

// ============================================================================
// THE CROSS-LANGUAGE CONTRACT (docs/determinism.md).
// This function is mirrored byte-for-byte in behavior by
// client/CoreLogic/MovementKernel.cs and pinned by shared/golden/
// movement_vectors.json. Any change here REQUIRES regenerating the golden
// vectors (`npm run golden`) and porting the change to the C# side.
// ============================================================================

export interface KernelInput {
  moveX: number; // world-space, milli-normalized (-1000..1000)
  moveY: number;
}

export interface KernelTransform {
  x: number; // mu
  y: number; // mu
}

const MIN_MU = 50; // keep hitbox centers off the exact arena border

/**
 * Advance a possessed unit by one 50ms tick. Movement input is world-space
 * (the client resolves joystick+camera into world axes with local floats;
 * only the resulting integer vector enters the deterministic domain).
 * Axis-separated collision: try X then Y so walls slide instead of stick.
 */
export function stepPossessedMovement(
  input: KernelInput,
  t: KernelTransform,
  speedMuPerTick: number,
): KernelTransform {
  let mx = input.moveX;
  let my = input.moveY;

  // Clamp magnitude to 1000 so diagonal input is not faster.
  const mag = vecLen(mx, my);
  if (mag > 1000) {
    mx = idiv(mx * 1000, mag);
    my = idiv(my * 1000, mag);
  }

  const dx = idiv(mx * speedMuPerTick, 1000);
  const dy = idiv(my * speedMuPerTick, 1000);

  let nx = t.x;
  let ny = t.y;

  const candX = clampMu(nx + dx, GRID_W);
  if (isWalkableCell(idiv(candX, 1000), idiv(ny, 1000))) nx = candX;

  const candY = clampMu(ny + dy, GRID_H);
  if (isWalkableCell(idiv(nx, 1000), idiv(candY, 1000))) ny = candY;

  return { x: nx, y: ny };
}

function clampMu(v: number, cells: number): number {
  const max = cells * 1000 - MIN_MU;
  return v < MIN_MU ? MIN_MU : v > max ? max : v;
}

File: client/CoreLogic/MovementKernel.cs
namespace Overlord.CoreLogic;

public readonly record struct KernelInput(int MoveX, int MoveY);

public readonly record struct KernelTransform(int X, int Y);

/// <summary>
/// THE CROSS-LANGUAGE CONTRACT — bit-exact mirror of
/// packages/sim/src/movementKernel.ts, pinned by
/// shared/golden/movement_vectors.json (MovementKernelGoldenTests).
/// Any change must be made on the TS side first, golden vectors regenerated
/// (npm run golden), and ported here verbatim.
/// </summary>
public static class MovementKernel
{
    private const int MinMu = 50;

    public static KernelTransform StepPossessedMovement(
        KernelInput input,
        KernelTransform t,
        int speedMuPerTick)
    {
        var mx = input.MoveX;
        var my = input.MoveY;

        var mag = FixedMath.VecLen(mx, my);
        if (mag > 1000)
        {
            mx = mx * 1000 / mag;
            my = my * 1000 / mag;
        }

        var dx = mx * speedMuPerTick / 1000;
        var dy = my * speedMuPerTick / 1000;

        var nx = t.X;
        var ny = t.Y;

        var candX = ClampMu(nx + dx, GridModel.GridW);
        if (GridModel.IsWalkableCell(candX / 1000, ny / 1000)) nx = candX;

        var candY = ClampMu(ny + dy, GridModel.GridH);
        if (GridModel.IsWalkableCell(nx / 1000, candY / 1000)) ny = candY;

        return new KernelTransform(nx, ny);
    }

    private static int ClampMu(int v, int cells)
    {
        var max = cells * 1000 - MinMu;
        return v < MinMu ? MinMu : v > max ? max : v;
    }
}
